File: src/func/lex.py
class TextAnalysis:
    @staticmethod
    def ishexdigit(c):
        return c.isdigit() or (ord(c) >= ord('a') and ord(c) <= ord('f'))

class Token:
    def __init__(self, tp, s):
        self.tp = tp
        self.s = s

    def __repr__(self):
        return f"Token({self.tp}, {self.s})"

    def __str__(self):
        return repr(self)

class LexIter:
    class NullTerminatedString:
        def __init__(self, s):
            self.s = s
        def __getitem__(self, key):
            if key >= 0 and key < len(self.s):
                return self.s[key]
            return "\0"

    def __init__(self, s):
        self.s = LexIter.NullTerminatedString(s)
        self.pos = 0
        self.done = False

    def __next__(self):
        if (self.done):
            raise StopIteration

        s = self.s
        while s[self.pos].isspace():
            self.pos += 1
        
        # check actual character
        single_chars = {
                '>' : "mover", '<' : "movel", '=' : "print" , 
                '{' : "lcurly", '}' : "rcurly",
                '(' : "lpar", ')' : "rpar",
                '[' : "lbra", ']' : "rbra",
                ',' : "comma",
                '$' : "chr_var"
        }
        c = s[self.pos]
        if c == '\0':
            self.done = True
            return Token("eof", None)
        if c in single_chars:
            self.pos += 1
            return Token(single_chars[c], c)
        if c == '\'':
            c1 = s[self.pos + 1]
            c2 = s[self.pos + 2]
            if c1 == 'x':
                if c2 == '\'':
                    self.pos += 3
                    return Token("chr_imm", "x")
                c3 = s[self.pos + 3]
                c4 = s[self.pos + 4]
                if TextAnalysis.ishexdigit(c2) and TextAnalysis.ishexdigit(c3) and c4 == '\'':
                    self.pos += 5
                    return Token("chr_imm", "x" + c2 + c3)
                self.done = True
                return Token("error", None)
            if ord(c1) >= ord('!') and ord(c1) <= ord('~') and c2 == '\'':
                self.pos += 3
                return Token("chr_imm", c1)
            self.done = True
            return Token("error", None)
        if c.isalpha() or c == '_':
            start = self.pos
            self.pos += 1
            while s[self.pos].isalnum() or s[self.pos] == '_':
                self.pos += 1
            return Token("ident", s.s[start:self.pos])
        self.done = True
        return Token("error", None)

class Lexer:
    class NullTerminatedString:
        def __init__(self, s):
            self.s = s
        def __getitem__(self, key):
            if isinstance(key, slice) or (key >= 0 and key < len(self.s)):
                return self.s[key]
            return "\0"

    def __init__(self, s):
        self.s = Lexer.NullTerminatedString(s)
        self.pos = 0
        self.tok = None
        self.done = False
        self.lex()

    def reset(self):
        self.pos = 0
        self.done = False
        self.lex()

    def _lex(self):
        if self.done:
            return

        s = self.s
        while s[self.pos].isspace():
            self.pos += 1
        
        # check actual character
        single_chars = {
                '>' : "mover", '<' : "movel", '=' : "print" , 
                '{' : "lcurly", '}' : "rcurly",
                '(' : "lpar", ')' : "rpar",
                '[' : "lbra", ']' : "rbra",
                ',' : "comma",
                '$' : "chr_var"
        }
        c = s[self.pos]
        if c == '\0':
            self.done = True
            self.tok = Token("eof", None)
            return
        if c in single_chars:
            self.pos += 1
            self.tok = Token(single_chars[c], c)
            return
        if c == '\'':
            c1 = s[self.pos + 1]
            c2 = s[self.pos + 2]
            if c1 == 'x':
                if c2 == '\'':
                    self.pos += 3
                    self.tok = Token("chr_imm", "x")
                    return
                c3 = s[self.pos + 3]
                c4 = s[self.pos + 4]
                if TextAnalysis.ishexdigit(c2) and TextAnalysis.ishexdigit(c3) and c4 == '\'':
                    self.pos += 5
                    self.tok = Token("chr_imm", "x" + c2 + c3)
                    return
                self.done = True
                self.tok = Token("error", None)
                return
            if ord(c1) >= ord('!') and ord(c1) <= ord('~') and c2 == '\'':
                self.pos += 3
                self.tok = Token("chr_imm", c1)
                return
            self.done = True
            self.tok = Token("error", None)
            return
        if c.isalpha() or c == '_':
            start = self.pos
            self.pos += 1
            while s[self.pos].isalnum() or s[self.pos] == '_':
                self.pos += 1
            self.tok = Token("ident", s.s[start:self.pos])
            return
        self.done = True
        self.tok = Token("error", None)

    def lex(self):
        self._lex()
        print(f"lex: {self.tok}")

if __name__ == "__main__":
    t = """
      fr($a, E) {
        [$a] E
        ['f'] <<='f'<< main
        ['x00', 'x0f'] _Other2
        fr($a, E)
      }
    """
    lx = Lexer(t)
    print(lx.tok)
    while lx.tok.tp not in ("eof", "error"):
        lx.lex()
        print(lx.tok)

File: src/func/lex.cpp
#include <cctype>
#include <stdexcept>

#include "lex.hpp"

using std::string;
using std::runtime_error;

namespace lex {

string Lexer::substring() const {
    if (type == eof) {
        return "";
    }
    return s.substr(offset, len);
}

void Lexer::lex() {
    if (done)
        return;

    while(isspace(s[cur]))
        ++cur;

    string single_chars("<>={}()[],$");
    char c = s[cur];
    if (c == '\0') {
        done = true;
        settoken(eof, 1, cur);
        return;
    } else if (single_chars.find(c) != string::npos) {
        settoken((TokenType) c, 1, cur);
        ++cur;
        return;
    } else if (c == '\'') {
        char c1 = s[cur + 1];
        if (c1 == 'x') {
            char c2 = s[cur + 2];
            if (c2 == '\'') {
                settoken(chr_imm, 1, cur + 1);
                cur += 3;
                return;
            } else if (isxdigit(c2)) {
                char c3 = s[cur + 3];
                if (isxdigit(c3)) {
                    char c4 = s[cur + 4];
                    if (c4 == '\'') {
                        settoken(chr_imm, 3, cur + 1);
                        cur += 5;
                        return;
                    }
                }
            }
        } else if (c1 >= '!' && c1 <= '~') {
            char c2 = s[cur + 2];
            if (c2 == '\'') {
                settoken(chr_imm, 1, cur + 1);
                cur += 3;
                return;
            }
        }
    } else if (isalpha(c) || c == '_') {
        auto start = cur;
        ++cur;
        while (isalnum(s[cur]) || s[cur] == '_')
            ++cur;
        settoken(ident, cur - start, start);
        return;
    }

    done = true;
    settoken(error, 1, cur);
}

string Lexer::tokname(TokenType t) {
    static string names[] = {
        "eof", "error", "chr_imm", "ident",
    };
    if (t <= 3) {
        return names[t];
    }
    return string(1, t);
}

void Lexer::expect(TokenType t) {
    if (type != t)
        throw runtime_error("expected " + tokname(t) + " but got " + tokname(type));
    lex();
}

}
